core-js/src/definitions.ts

```diff
@@ -0,0 +1,1421 @@
+import {
+  ASTNode,
+  DirectiveLocation,
+  DirectiveLocationEnum,
+  GraphQLError
+} from "graphql";
+import { assert } from "./utils";
+import deepEqual from 'deep-equal';
+
+export type QueryRoot = 'query';
+export type MutationRoot = 'mutation';
+export type SubscriptionRoot = 'subscription';
+export type SchemaRoot = QueryRoot | MutationRoot | SubscriptionRoot;
+
+export type Type = InputType | OutputType;
+export type NamedType = ScalarType | ObjectType | InterfaceType | UnionType | EnumType | InputObjectType;
+export type OutputType = ScalarType | ObjectType | InterfaceType | UnionType | EnumType | ListType<any> | NonNullType<any>;
+export type InputType = ScalarType | EnumType | InputObjectType | ListType<any> | NonNullType<any>;
+export type WrapperType = ListType<any> | NonNullType<any>;
+
+export type OutputTypeReferencer = FieldDefinition<any>;
+export type InputTypeReferencer = InputFieldDefinition | ArgumentDefinition<any>;
+export type ObjectTypeReferencer = OutputTypeReferencer | UnionType | SchemaDefinition;
+export type InterfaceTypeReferencer = OutputTypeReferencer | ObjectType | InterfaceType;
+
+export type NullableType = NamedType | ListType<any>;
+
+export type NamedTypeKind = NamedType['kind'];
+
+export function defaultRootTypeName(root: SchemaRoot) {
+  return root.charAt(0).toUpperCase() + root.slice(1);
+}
+
+export function isNamedType(type: Type): type is NamedType {
+  return type instanceof BaseNamedType;
+}
+
+export function isWrapperType(type: Type): type is WrapperType {
+  return type.kind == 'ListType' || type.kind == 'NonNullType';
+}
+
+export function isOutputType(type: Type): type is OutputType {
+  if (isWrapperType(type)) {
+    return isOutputType(type.baseType());
+  }
+  switch (type.kind) {
+    case 'ScalarType':
+    case 'ObjectType':
+    case 'UnionType':
+    case 'EnumType':
+    case 'InterfaceType':
+      return true;
+    default:
+      return false;
+  }
+}
+
+export function ensureOutputType(type: Type): OutputType {
+  if (isOutputType(type)) {
+    return type;
+  } else {
+    throw new Error(`Type ${type} (${type.kind}) is not an output type`);
+  }
+}
+
+export function isInputType(type: Type): type is InputType {
+  if (isWrapperType(type)) {
+    return isInputType(type.baseType());
+  }
+  switch (type.kind) {
+    case 'ScalarType':
+    case 'EnumType':
+    case 'InputObjectType':
+      return true;
+    default:
+      return false;
+  }
+}
+
+export function ensureInputType(type: Type): InputType {
+  if (isInputType(type)) {
+    return type;
+  } else {
+    throw new Error(`Type ${type} (${type.kind}) is not an input type`);
+  }
+}
+
+export interface Named {
+  readonly name: string;
+}
+
+export abstract class SchemaElement<Parent extends SchemaElement<any, any> | Schema, Referencer> {
+  protected _parent?: Parent;
+  protected readonly _appliedDirectives: Directive[] = [];
+  description?: string;
+  source?: ASTNode;
```

Comment:
Initially I was planning to generalise this somehow to a more general notion of "sources". At least, for composition, it would made sense for a given element to point back to each subgraph nodes that were merged into that particular element (so in that case, the source would have to be more like a map of subgraph name to `ASTNode`).

I also wanted to carry the notion that this `ASTNode` might not represent the state of the `Schema` anymore, it's only the state of when it was created. And to clarify, what I mean here is that if you consider an `ObjectType` for instance, the `source` will properly represent the type after parsing, but if you modify that type now (say, add a field), then the code is not clearly the `source` (because it feels useful to still being able to point to the original source for errors) but that source is not in sync with the object anymore.

Anyway, I'll probably rename to something like `sourceAST` for now. But I might generalise this soonish when I work on composition.